src/bin/solang.rs

```diff
@@ -470,6 +470,11 @@ fn process_file(
             eprintln!("{}: error: {}", dot_filename.display(), err);
             std::process::exit(1);
         }
+
+        if ns.contracts.is_empty() || diagnostics::any_errors(&ns.diagnostics) {
+            return Err(());
+        }
+
```

Comment:
That is not the right solution, as if when you run `solang --target solana --emit ast-dot foo.sol` and foo.sol is valid, a binary file is still created (`bundle.so` and abi files)

Reply:
Does this problem only happen with the Solana target?

Reply:
I think so, you could try it out.

Reply:
Sorry, I think I am mistaken. This does not happen on solana either.

However, when you run `solang --target solana --emit ast-doc foo.sol` and there are compiler errors, then you both get a dot file and the message `error: not all contracts are valid` and non-zero exit code. This suggests that the command did not work, but it did and you do have a .dot file

Reply:
I think the issue is that `solang::compile_many()` is called when there are compiler errors for the solana target

Reply:
I think that the issue is that we don't return errors from the function `process_file` when we emit the `ast-dot`.
Have a look here:
https://github.com/hyperledger-labs/solang/blob/a524ff58f5cc033b849195c36cb0e07e06fe05fb/src/bin/solang.rs#L473
As we always return `Ok`, we will never exit the process on the following condition:
https://github.com/hyperledger-labs/solang/blob/a524ff58f5cc033b849195c36cb0e07e06fe05fb/src/bin/solang.rs#L336

Reply:
When I add the checks above, I return an error from `process_file` and the `solang::compile_many()` won't be called.

Reply:
The problem with that is that the behaviour of the solang command line is a bit in-consistent with your change:

```
[sean@xywoleh solang]$ cargo run foo.sol --target solana --emit ast-dot -v
   Compiling solang v0.1.10 (/home/sean/git/solang)
    Finished dev [unoptimized + debuginfo] target(s) in 26.59s
     Running `target/debug/solang foo.sol --target solana --emit ast-dot -v`
info: Solang version v0.1.10-20-g51904a31
help: found contract ‘c’
  ┌─ /home/sean/git/solang/foo.sol:1:1
  │
1 │ contract c {
  │ ^^^^^^^^^^^

error: function cannot have same name as the contract
  ┌─ /home/sean/git/solang/foo.sol:2:2
  │
2 │     function c() public {}
  │     ^^^^^^^^^^^^^^^^^^^

info: Saving graphviz dot ./foo.dot
error: not all contracts are valid
[sean@xywoleh solang]$ echo $?
1
```

So now it returns an error exit code but it did actually save the dot file. That is not behaviour I would expect. Without the `-v` option it is even worse: you have no idea that it succeeded in writing the dot file.

Reply:
What do you suggest then?

Reply:
Change line 346 to:
```
if target == solang::Target::Solana && !diagnostics::any_errors(&ns.diagnostics) {
```